### PR TITLE
Fix a problem where the newly provided mode is a string

### LINKED ISSUE #3597: Unarchive Module fails with Traceback
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

unarchive
##### ANSIBLE VERSION

```
[root@8a976966193f stuff]# ansible --version
ansible 2.2.0 (devel cf62a62b83) last updated 2016/05/03 23:19:15 (GMT +000)
  lib/ansible/modules/core: (detached HEAD 9db1233521) last updated 2016/05/03 23:19:24 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 675d778b50) last updated 2016/05/03 23:19:29 (GMT +000)
  config file = /stuff/ansible.cfg
  configured module search path = ['./library']
```
##### OS / ENVIRONMENT

centos based Docker image
##### SUMMARY

While using the unarchive module to unzip consul from a /tmp directory on the remote machine to the destination /usr/local/bin the module throws an exception. In this case, consul had previously been installed, so /usr/local/bin/consul already existed.

```
fatal: [10.64.1.198]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_5Ccjgl/ansible_module_unarchive.py\", line 751, in <module>\r\n    main()\r\n  File \"/tmp/ansible_5Ccjgl/ansible_module_unarchive.py\", line 711, in main\r\n    check_results = handler.is_unarchived()\r\n  File \"/tmp/ansible_5Ccjgl/ansible_module_unarchive.py\", line 397, in is_unarchived\r\n    err += 'Path %s differs in permissions (%o vs %o)\\n' % (path, self.file_args['mode'], stat.S_IMODE(st.st_mode))\r\nTypeError: %o format: a number is required, not str\r\n", "msg": "MODULE FAILURE", "parsed": false}
```
##### STEPS TO REPRODUCE

The tasks I'm running are... (my role is not very different from https://github.com/savagegus/ansible-consul)

```
- name: downoad consul
  get_url: >
    url={{ consul_release_url }}
    checksum={{ consul_release_checksum }}
    dest={{ consul_download_folder }}/{{ consul_archive }}
  register: consul_downloaded
  become: yes

- name: install consul
  unarchive: >
    src={{ consul_download_folder }}/{{ consul_archive }}
    dest=/usr/local/bin
    owner={{ consul_user }}
    group={{ consul_group }}
    mode=755
    copy=no
  become: yes
```

##### EXPECTED RESULTS

Previously with the current stable release of ansible (ansible 2.0.2) this task runs without any errors. 
##### ACTUAL RESULTS

I can run with higher verbosity if you think its helpful. 

```
fatal: [10.64.1.78]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_bGfFXq/ansible_module_unarchive.py\", line 751, in <module>\r\n    main()\r\n  File \"/tmp/ansible_bGfFXq/ansible_module_unarchive.py\", line 711, in main\r\n    check_results = handler.is_unarchived()\r\n  File \"/tmp/ansible_bGfFXq/ansible_module_unarchive.py\", line 397, in is_unarchived\r\n    err += 'Path %s differs in permissions (%o vs %o)\\n' % (path, self.file_args['mode'], stat.S_IMODE(st.st_mode))\r\nTypeError: %o format: a number is required, not str\r\n", "msg": "MODULE FAILURE", "parsed": false}
```

I worked around this by changing the string formatting thats throwing the error to cast everything as a string. Probably not the correct way to do it which is why I'm not creating a pull request, but it gets me past the error for now. 

```
--- a/files/unarchive.py
+++ b/files/unarchive.py
@@ -394,7 +394,7 @@ class ZipArchive(object):
                     mode = self._permstr_to_octal(permstr, umask)
                 if self.file_args['mode'] and  self.file_args['mode'] != stat.S_IMODE(st.st_mode):
                     change = True
-                    err += 'Path %s differs in permissions (%o vs %o)\n' % (path, self.file_args['mode'], stat.S_IMODE(st.st_mode))
+                    err += 'Path %s differs in permissions (%s vs %s)\n' % (path, str(self.file_args['mode']), str(stat.S_IMODE(st.st_
                     itemized[5] = 'p'
                 elif mode != stat.S_IMODE(st.st_mode):
                     change = True
```

### PR DESCRIPTION
**Pull request by dagwieers**
##### ISSUE TYPE

- Bugfix Pull Request
##### COMPONENT NAME

unarchive
##### SUMMARY

This fixes #3597

### CODE REVIEW

**Comment by dagwieers** [comment]
@tkinz27 Please test !

**Comment by gregdek** [comment]
Thanks again to @dagwieers. This module is going into community review.

We notice that you are one of the maintainers (@dagwieers, @pileofrogs) of this module, so when you think it's ready, please comment "shipit" and we will consider it for merging.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @dagwieers for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by abadger** [comment]
Note: I briefly looked at this yesterday.  Got approximately this far too
but then realized mode can also be symbolic (example: "u+rX,go-w". )

**Comment by gregdek** [comment]
@dagwieers A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@dagwieers Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by dagwieers** [comment]
I am not going to implement symbolic mode support, so either don't add this (and leave the issue in place until you have a complete fix) or merge it to get a partial fix. @abadger

File: files/unarchive.py [hunk 1]
```diff
@@ -430,16 +430,23 @@ def is_unarchived(self):
...
             if ftype != 'L':
+
+                # Use the new mode provided with the action, if there is one
+                if self.file_args['mode']:
+                    if isinstance(self.file_args['mode'], int):
+                        mode = self.file_args['mode']
+                    else:
+                        try:
+                            mode = int(self.file_args['mode'], 8)
+                        except Exception, e:
```
Comment (abadger) [inline files/unarchive.py]: Need to change this to be pyhton3 compatible as well:

``` python
[.. somewhere near the top ...]
from ansible.module_utils.pycompat24 import get_exception
[...]
except Exception:
    e = get_exception()
```

**Comment by abadger** [comment]
Okay, one thing to fix so that the travis tests pass.  I'll merge this -- and then we should treat the documentation (which mentions symbolic mode being supported) as the bg rather than the code.

**Comment by dagwieers** [comment]
Right, I added the get_exception() compatibility clause for python3.

**Comment by abadger** [comment]
Merged to devel and cherry-picked to stable-2.1

**Pull request merged**